In [170]:
import pandas as pd
import numpy as np
import re

df = pd.read_csv('D:\\BIG DATA LEVEL-2\\WEB ANALYTICS AND BUSINESS INTELLIGENCE\\FIFA 21.csv',low_memory=False)
print(df)

           ID               Name                      LongName  \
0      158023           L. Messi                  Lionel Messi   
1       20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2      200389           J. Oblak                     Jan Oblak   
3      192985       K. De Bruyne               Kevin De Bruyne   
4      190871          Neymar Jr    Neymar da Silva Santos Jr.   
...       ...                ...                           ...   
18974  247223             Xia Ao                        Ao Xia   
18975  258760           B. Hough                     Ben Hough   
18976  252757        R. McKinley                Ronan McKinley   
18977  243790       Wang Zhen'ao                  Zhen'ao Wang   
18978  252520          Zhou Xiao                     Xiao Zhou   

                                               photoUrl  \
0      https://cdn.sofifa.com/players/158/023/21_60.png   
1      https://cdn.sofifa.com/players/020/801/21_60.png   
2      https://cdn.sofifa.com/

In [171]:
df.head(5)

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club  ...     A/W     D/W   IR PAC SHO  PAS  \
0   93         \n\n\n\nFC Barcelona  ...  Medium     Low  5 ★  85  92   91   
1   92             \n\n\n\nJuventus  ...    High     Low  5 ★  89  93   81   
2   93      \n\n\n\nAtlético Madrid  ...  Medium  Medium  3 ★  87  92   78   
3   91      \n\n\n\nManchester City  ...    High    High  4 ★  76  86   93   
4   91  \n\n\n\nParis Saint-Germain  ...    High  Medium  5 ★  91  85   86   

  DRI DEF PHY Hits  
0  95  38  65  771  
1  89  35  77  562  
2  90  52  90  150  
3  88  64  78  207  
4  94  36  59  595  

[5 rows x 77 columns]

In [172]:
df.isnull().sum()

ID              0
Name            0
LongName        0
photoUrl        0
playerUrl       0
             ... 
PAS             0
DRI             0
DEF             0
PHY             0
Hits         2595
Length: 77, dtype: int64

NOTE: Checking null values in dataset. We got to know that Hits column has 2595 null rows.

In [173]:
df.drop_duplicates
df.shape

(18979, 77)

In [174]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

Convert the height and weight columns to numerical forms.

In [175]:
height_counts = df['Height'].unique()
print(height_counts)

['170cm' '187cm' '188cm' '181cm' '175cm' '184cm' '191cm' '178cm' '193cm'
 '185cm' '199cm' '173cm' '168cm' '176cm' '177cm' '183cm' '180cm' '189cm'
 '179cm' '195cm' '172cm' '182cm' '186cm' '192cm' '165cm' '194cm' '167cm'
 '196cm' '163cm' '190cm' '174cm' '169cm' '171cm' '197cm' '200cm' '166cm'
 '6\'2"' '164cm' '198cm' '6\'3"' '6\'5"' '5\'11"' '6\'4"' '6\'1"' '6\'0"'
 '5\'10"' '5\'9"' '5\'6"' '5\'7"' '5\'4"' '201cm' '158cm' '162cm' '161cm'
 '160cm' '203cm' '157cm' '156cm' '202cm' '159cm' '206cm' '155cm']


In [176]:
def convert_height(height:str):
    if "cm" in height:
        return int(height.replace('cm',''))
    elif "'" in height:
        feet_inch = re.split("\'", height)
        return int(round(float(feet_inch[0])* 30.48 + float(feet_inch[1].replace('"', ''))* 2.54))

df['Height'] = df['Height'].apply(convert_height)

height_counts = df['Height'].value_counts()
print(height_counts)


Height
180    1480
178    1254
185    1185
183    1156
175    1094
188     935
182     887
184     817
186     816
177     738
187     711
181     708
176     670
179     662
173     610
190     583
174     540
170     498
191     466
189     462
172     458
193     322
192     298
171     277
168     218
169     178
194     177
196     142
195     138
167     114
165      94
166      63
197      58
198      41
164      31
163      20
199      17
162      13
201      13
200       7
160       7
161       6
202       4
158       3
203       2
159       2
157       1
156       1
206       1
155       1
Name: count, dtype: int64


NOTE: Here, converting all data into cm, and converting it to int format. Also remove unit cm from dataframe.

In [177]:
df.Height.dtypes

dtype('int64')

NOTE: Datatype of Heights is int now.

In [178]:
weight_counts = df['Weight'].unique()
print(weight_counts)

['72kg' '83kg' '87kg' '70kg' '68kg' '80kg' '71kg' '91kg' '73kg' '85kg'
 '92kg' '69kg' '84kg' '96kg' '81kg' '82kg' '75kg' '86kg' '89kg' '74kg'
 '76kg' '64kg' '78kg' '90kg' '66kg' '60kg' '94kg' '79kg' '67kg' '65kg'
 '59kg' '61kg' '93kg' '88kg' '97kg' '77kg' '62kg' '63kg' '95kg' '100kg'
 '58kg' '183lbs' '179lbs' '172lbs' '196lbs' '176lbs' '185lbs' '170lbs'
 '203lbs' '168lbs' '161lbs' '146lbs' '130lbs' '190lbs' '174lbs' '148lbs'
 '165lbs' '159lbs' '192lbs' '181lbs' '139lbs' '154lbs' '157lbs' '163lbs'
 '98kg' '103kg' '99kg' '102kg' '56kg' '101kg' '57kg' '55kg' '104kg'
 '107kg' '110kg' '53kg' '50kg' '54kg' '52kg']


In [179]:
def convert_weight(weight: str):
    if "kg" in weight:
        return int(weight.replace('kg',''))
    elif "lb" in weight:
        lbs = int(weight.replace('lbs',''))
        return int(round(lbs*0.453592))

df['Weight'] = df['Weight'].apply(convert_weight)

weight_counts = df['Weight'].value_counts()
print(weight_counts)

Weight
70     1496
75     1459
80     1109
72     1024
78      995
74      957
73      956
77      893
76      880
68      795
79      684
71      648
82      622
69      565
65      535
81      523
83      515
67      508
84      444
85      423
66      401
64      341
86      285
88      266
63      239
87      224
62      181
90      156
60      153
89      128
61       96
91       87
92       68
93       52
58       44
59       41
94       34
95       30
57       19
55       17
96       15
98       13
56       12
97        9
100       8
54        5
103       4
102       4
101       4
99        3
104       3
52        2
107       1
110       1
53        1
50        1
Name: count, dtype: int64


NOTE: Weight should be convert in kg and int datatype. And weight which are in lbs is rounded with a formula.

Remove the unnecessary newline characters from all columns that have them. 

In [180]:
columns_with_newline = []
for column in df.columns:
    if df[column].apply(lambda x: '\n' in str(x)).any():
        columns_with_newline.append(column)

# Print columns with newline characters
print("Columns with newline characters:", columns_with_newline)

Columns with newline characters: ['Club']


In [181]:
df['Club'] = df['Club'].str.replace('\n', '', regex=True)
print(df['Club'].head(5))

0           FC Barcelona
1               Juventus
2        Atlético Madrid
3        Manchester City
4    Paris Saint-Germain
Name: Club, dtype: object


NOTE: Column name 'Club' has new line character in the beginning. By replacing \n with '' it removes new line character.

'Value', 'Wage' and "Release Clause' are string columns. Convert them to numbers. For eg, "M" in value column is Million, so multiply the row values by 1,000,000, etc. 

In [182]:
def convert_money(money: str):
    if '€' in money:
        money=money.replace("€","")
    if 'M' in money:
        return int(float(money.replace('M',''))*1000000)
    elif 'K' in money:
        return int(float(money.replace('K',''))*1000)
    else:
        return int(money) 

df['Value'] = df['Value'].apply(convert_money)
df['Wage'] = df['Wage'].apply(convert_money)
df['Release Clause'] = df['Release Clause'].apply(convert_money)
df[['Value','Wage','Release Clause']]

Value    Wage  Release Clause
0      103500000  560000       138400000
1       63000000  220000        75900000
2      120000000  125000       159400000
3      129000000  370000       161000000
4      132000000  270000       166500000
...          ...     ...             ...
18974     100000    1000           70000
18975     130000     500          165000
18976     120000     500          131000
18977     100000    2000           88000
18978     100000    1000           79000

[18979 rows x 3 columns]

NOTE: Covert_money function replace € with "", number which contain M multiply with 1000000 and K with 1000.

Some columns have 'star' characters. Strip those columns of these stars and make the columns numerical.

In [183]:
columns_with_star = []
for column in df.columns:
    if df[column].apply(lambda x: '★' in str(x)).any():
        columns_with_star.append(column)

# Print columns with star characters
print("Columns with star characters:", columns_with_star)

Columns with star characters: ['W/F', 'SM', 'IR']


In [184]:
columns_star = ['W/F', 'SM', 'IR']
for column in columns_star:
    df[column] = df[column].replace('★', '', regex=True)
print(df[columns_star])

      W/F SM  IR
0      4   4  5 
1      4   5  5 
2      3   1  3 
3      5   4  4 
4      5   5  5 
...    .. ..  ..
18974  2   2  1 
18975  2   2  1 
18976  2   2  1 
18977  3   2  1 
18978  3   2  1 

[18979 rows x 3 columns]


NOTE: Column names W/F,SM,IR columns has ★. So, here replacing that with space. By using lambda function to create simple expression and call the function. 

Remove further outliers from data and generate a summary report stating all the changes.

In [185]:
df = df.rename(columns={'LongName':'FullName'})

NOTE: Renaming LongName with FullName.

In [186]:
df.Hits.unique()

array(['771', '562', '150', '207', '595', '248', '246', '120', '1.6K',
       '130', '321', '189', '175', '96', '118', '216', '212', '154',
       '205', '202', '339', '408', '103', '332', '86', '173', '161',
       '396', '1.1K', '433', '242', '206', '177', '1.5K', '198', '459',
       '117', '119', '209', '84', '187', '165', '203', '65', '336', '126',
       '313', '124', '145', '538', '182', '101', '45', '377', '99', '194',
       '403', '414', '593', '374', '245', '3.2K', '266', '299', '309',
       '215', '265', '211', '112', '337', '70', '159', '688', '116', '63',
       '144', '123', '71', '224', '113', '168', '61', '89', '137', '278',
       '75', '148', '176', '197', '264', '214', '247', '402', '440',
       '1.7K', '2.3K', '171', '320', '657', '87', '259', '200', '255',
       '253', '196', '60', '97', '85', '169', '256', '132', '239', '166',
       '121', '109', '32', '46', '122', '48', '527', '199', '282', '51',
       '1.9K', '642', '155', '323', '288', '497', '509', '79',

In [187]:
df.Hits = df.Hits.fillna('0')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   FullName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  int64 
 13  Weight            18979 non-null  int64 
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

NOTE: There are some data in Hits column are null. Therefore, filling that rows with 0.

In [188]:
def covert_Hits(hits: str):
    hits = str(hits)
    if 'K' not in hits:
        return float(hits)
    else:
        return float(hits.replace('K',''))*1000
        
df['Hits'] = df['Hits'].apply(covert_Hits)
df['Hits']

0        771.0
1        562.0
2        150.0
3        207.0
4        595.0
         ...  
18974      0.0
18975      0.0
18976      0.0
18977      0.0
18978      0.0
Name: Hits, Length: 18979, dtype: float64

NOTE: Here, column Hits has K means thousand so, multiplying this with 1000.

In [189]:
df = df.rename(columns={'↓OVA':'OVA'})

NOTE: Renaming ↓OVA with OVA.

In [190]:
from datetime import datetime
df['Joined'] = df['Joined'].apply(lambda x: datetime.strptime(x, "%d-%b-%y"))

In [191]:
df['Joined']

0       2004-07-01
1       2018-07-10
2       2014-07-16
3       2015-08-30
4       2017-08-03
           ...    
18974   2018-07-13
18975   2020-08-01
18976   2019-03-08
18977   2020-09-22
18978   2019-07-29
Name: Joined, Length: 18979, dtype: datetime64[ns]

NOTE: By using strptime() method creates a datetime object from the string in this format "%d-%b-%y".

In [192]:
df.Contract.unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

In [193]:
def Clean_Contract(a):
    if '~' in a:
        start_year, end_year = a.split(' ~ ')
        return int(end_year) - int(start_year)
    else:
        return 0


In [194]:
df['Contract_Duration'] = df['Contract'].apply(Clean_Contract)
# Rename the column to Contract status
df.Contract_Duration.head()

0    17
1     4
2     9
3     8
4     5
Name: Contract_Duration, dtype: int64

NOTE: Clean_Contract function refers to split years which has '~' and provide difference of it for cleaning of Contract column and made new column name Contract_Duration.

In [195]:
df.Contract_Duration.unique()

array([17,  4,  9,  8,  5,  6,  7, 10, 12, 13, 16, 14, 15,  2, 11,  3,  1,
        0, 20, 18, 21, 19, 23], dtype=int64)

In [196]:
df['Positions'].unique()

array(['RW, ST, CF', 'ST, LW', 'GK', 'CAM, CM', 'LW, CAM', 'ST', 'RW',
       'ST, LW, RW', 'CB', 'LW', 'CDM', 'CF, ST', 'LW, RW', 'CDM, CM',
       'CDM, RB', 'CF, CAM', 'LW, ST', 'CM', 'ST, CF, LW', 'RM, LM, CAM',
       'RB', 'RW, CAM, CM', 'LB', 'LM, CF', 'CF', 'RW, LW', 'CAM, RM, RW',
       'CM, CDM', 'CAM, CF, ST', 'CM, CDM, CAM', 'CF, LW, CAM',
       'CAM, RM, CF', 'LM, ST', 'RM, LM, RW', 'LM', 'CAM, RW', 'CB, CDM',
       'RW, RM', 'LW, CF', 'CM, RM, LM', 'LB, LM', 'CAM, CM, RM',
       'CAM, CM, CF', 'CAM, CF', 'LM, RM, LW', 'LM, LB, CM', 'CM, LM, LB',
       'RM, RW', 'RM, CM', 'CAM, CM, LW', 'CB, LB', 'RM, RB', 'ST, RW',
       'LM, RW, LW', 'RB, LB', 'RB, RM', 'RM', 'LM, RM, CF', 'CAM, RM',
       'RB, RWB', 'CDM, CB, CM', 'CAM, RM, ST', 'LM, LW, RM', 'CM, CAM',
       'ST, RM, CF', 'LM, RM', 'RM, CF', 'LM, LWB', 'RW, RM, CF',
       'RB, CM', 'LW, CAM, RW', 'CAM, LW, CM', 'CM, CAM, CDM',
       'RW, LW, CAM', 'CM, CAM, LM', 'CM, RM, ST', 'CDM, CM, RB',
       'ST, CAM', 

In [197]:
temp_position = []
for x in range(len(df['Positions'])):
    y = sorted(df['Positions'][x].split(" "))
    yx = ' '.join(y)
    temp_position.append(yx)

df['Positions'] = temp_position
df['Positions'] = df['Positions'].astype(object)
df['Positions'].unique()

array(['CF RW, ST,', 'LW ST,', 'GK', 'CAM, CM', 'CAM LW,', 'ST', 'RW',
       'LW, RW ST,', 'CB', 'LW', 'CDM', 'CF, ST', 'LW, RW', 'CDM, CM',
       'CDM, RB', 'CAM CF,', 'LW, ST', 'CM', 'CF, LW ST,', 'CAM LM, RM,',
       'RB', 'CAM, CM RW,', 'LB', 'CF LM,', 'CF', 'LW RW,', 'CAM, RM, RW',
       'CDM CM,', 'CAM, CF, ST', 'CAM CDM, CM,', 'CAM CF, LW,',
       'CAM, CF RM,', 'LM, ST', 'LM, RM, RW', 'LM', 'CAM, RW', 'CB, CDM',
       'RM RW,', 'CF LW,', 'CM, LM RM,', 'LB, LM', 'CAM, CM, RM',
       'CAM, CF CM,', 'CAM, CF', 'LM, LW RM,', 'CM LB, LM,', 'CM, LB LM,',
       'RM, RW', 'CM RM,', 'CAM, CM, LW', 'CB, LB', 'RB RM,', 'RW ST,',
       'LM, LW RW,', 'LB RB,', 'RB, RM', 'RM', 'CF LM, RM,', 'CAM, RM',
       'RB, RWB', 'CB, CDM, CM', 'CAM, RM, ST', 'LM, LW, RM', 'CAM CM,',
       'CF RM, ST,', 'LM, RM', 'CF RM,', 'LM, LWB', 'CF RM, RW,',
       'CM RB,', 'CAM, LW, RW', 'CAM, CM LW,', 'CAM, CDM CM,',
       'CAM LW, RW,', 'CAM, CM, LM', 'CM, RM, ST', 'CDM, CM, RB',
       'CAM ST,', 

NOTE: It sort the Positions column and split it.

In [198]:
Final_exam = 'D:/BIG DATA LEVEL-2/WEB ANALYTICS AND BUSINESS INTELLIGENCE/Drashti(8890040)_Final.xlsx'
df.to_excel(Final_exam)